tc/proto/compcache.proto

```diff
@@ -16,57 +16,72 @@ message TensorInfoProto {
   required DLDataTypeProto dtype = 4;
 }
 
-message CudaCacheEntryProto {
-  required string id = 1;
-  required CudaMappingOptionsProto kernel_options = 2;
-  repeated TensorInfoProto inputs = 3;
-  repeated TensorInfoProto outputs = 4;
-  required string device_str = 5;
-  required string git_version = 6;
-
-  required string cuda_source = 7;
-  required string specialized_name = 8;
-  repeated sint32 parameters = 9;
-  required CudaDimProto grid_dims = 10;
-  required CudaDimProto block_dims = 11;
-}
-
-message ManualCudaCacheEntryProto {
+message OptionsCacheKeyProto {
   required string id = 1;
   repeated TensorInfoProto inputs = 2;
   repeated TensorInfoProto outputs = 3;
   required string device_str = 4;
   required string git_version = 5;
-
-  required string cuda_source = 6;
-  repeated uint64 grid_dims = 7;
-  repeated uint64 block_dims = 8;
 }
 
-message OptionsCacheValuesProto{
+message CudaOptionsCacheValueProto{
   required CudaMappingOptionsProto kernel_options = 1;
   repeated uint64 recorded_runtimes = 2;
 }
 
-message OptionsCacheEntryProto {
-  required string id = 1;
-  repeated TensorInfoProto inputs = 2;
-  repeated TensorInfoProto outputs = 3;
-  required string device_str = 4;
-  required string git_version = 5;
+message CudaOptionsCacheProto {
+  repeated OptionsCacheKeyProto keys = 1;
```

Comment:
All these things should go away and we should use ```proto3``` with a simple map.

> the key and the value merged in one message and don't have this problem

I'd claim they have many other problems. Look in particular at the implementation of CudaCompilationCache which is currently distributed across [1](https://github.com/facebookresearch/TensorComprehensions/blob/master/tc/core/compilation_cache.h), [2](https://github.com/facebookresearch/TensorComprehensions/blob/master/tc/core/compilation_cache.cc), [3](https://github.com/facebookresearch/TensorComprehensions/blob/master/tc/core/cuda/cuda_compilation_cache-inl.h), [4](https://github.com/facebookresearch/TensorComprehensions/blob/master/tc/core/cuda/cuda_compilation_cache.cc), [5](https://github.com/facebookresearch/TensorComprehensions/blob/master/tc/core/cuda/cuda_compilation_cache.h) with CRTP fun.

Some of the complexity comes from specializing this data structure. But many problems come from trying to be too generic too early and all these templated searchKernel functions.

In this refactoring (and this OptionsCache RFC) I am going back to basics, backend templated multimap which matches the semantics of what an OptionsCache provides. We'll tend to problems when we hit them. All the rest is premature optimization which is now blowing to our faces when we try to be truly backend-agnostic.

One part of me really feel like ripping apart all these compilation caches as premature optimizations because in the new API we have the tuner with its cache (proposed here) and that's really the expensive part. The rest frankly could be tore down for now and redone from scratch. I already tore down the manual caches because in practice they are useless for now.

What would you think of tearing the CompilationCache apart for now?
Once autotuned, one would JIT-recompile anyway (and can reuse the same options for different input sizes) and can hold on to the TcExecutor.

cc @ttheodor

Reply:
And here is what a commit that drops compilation caches [looks like](https://github.com/nicolasvasilache/TensorComprehensions/commit/0f0d392012c6dfc6da75132935a25abfddc948bb) in the new API.

Compiles, tests pass etc.

Reply:
@nicolasvasilache  it doesn't sound like a bad idea. The cuda caches were used by the autotuner to avoid recompilations, however now that execution engine is dead (or will be soon) we can use the executors directly instead of relying on an underlying cache.

Reply:
I discussed with @math-fehr and it seems that compilation caches create overhead when we want to use cooperative groups (launching cooperative kernels uses a different call and adds extra overhead so we don't want to use it always).

So I would second the decision to remove the compilation caches.